src/ray/common/cgroup/cgroup_utils.cc

```diff
@@ -0,0 +1,185 @@
+    OpenCgroupV2FileAndAppend(max_memory_path, absl::StrFormat("%d", ctx.max_memory));
+  }
+
+  return true;
+}
+
+bool UpdateDefaultCgroupV2(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupV2Uuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupV2FilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  OpenCgroupV2FileAndAppend(procs_path, absl::StrFormat("%d", ctx.pid));
+
+  return true;
+}
+
+bool DeleteCgroupV2(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_NE(ctx.uuid, kDefaultCgroupV2Uuid);
+  RAY_CHECK_GT(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  return rmdir(cgroup_folder.data()) == 0;
+}
+
+bool RemoveCtxFromDefaultCgroupV2(const PhysicalModeExecutionContext &ctx) {
+  // Sanity check.
+  RAY_CHECK(!ctx.uuid.empty());
+  RAY_CHECK_EQ(ctx.uuid, kDefaultCgroupV2Uuid);
+  RAY_CHECK_EQ(ctx.max_memory, 0);
+
+  const std::string cgroup_folder =
+      absl::StrFormat("%s/%s", ctx.cgroup_directory, ctx.uuid);
+  int ret_code = mkdir(cgroup_folder.data(), kCgroupV2FilePerm);
+  if (ret_code != 0) {
+    return false;
+  }
+
+  const std::string procs_path = absl::StrFormat("%s/cgroup.procs", cgroup_folder);
+  std::ostringstream buffer;
+  {
+    std::ifstream file(procs_path.data(), std::ios::in);
+    buffer << file.rdbuf();
+  }
+  std::string content = buffer.str();  // contains all PIDs, separated by space
+
+  std::vector<std::string_view> old_pid_strings = absl::StrSplit(content, ' ');
+  std::vector<std::string_view> new_pid_strings;
+  new_pid_strings.reserve(old_pid_strings.size() - 1);
+  for (const auto &cur_pid : old_pid_strings) {
+    if (cur_pid == absl::StrFormat("%d", ctx.pid)) {
+      continue;
+    }
+    new_pid_strings.emplace_back(cur_pid);
+  }
+
+  const std::string new_pids = absl::StrJoin(new_pid_strings, " ");
+  {
+    std::ofstream out_file{procs_path.data(), std::ios::out};
+    out_file << new_pids;
+  }
+
+  return true;
+}
+
+}  // namespace
+
+/*static*/ std::unique_ptr<CgroupV2Setup> CgroupV2Setup::New(
+    PhysicalModeExecutionContext ctx) {
+  if (!CgroupV2Setup::SetupCgroupV2ForContext(ctx)) {
+    return nullptr;
+  }
+  return std::unique_ptr<CgroupV2Setup>(new CgroupV2Setup(std::move(ctx)));
+}
+
+CgroupV2Setup::~CgroupV2Setup() {
+  if (!CleanupCgroupV2ForContext(ctx_)) {
+    RAY_LOG(ERROR) << "Fails to cleanup cgroup for execution context with uuid "
+                   << ctx_.uuid;
+  }
+}
+
+/*static*/ bool CgroupV2Setup::SetupCgroupV2ForContext(
+    const PhysicalModeExecutionContext &ctx) {
+#ifndef __linux__
```

Comment:
these ifndef is getting messy. can we reorg code into:

```
#ifdef __linux__

bool CgroupV2Setup::SetupCgroupV2ForContext(
    const PhysicalModeExecutionContext &ctx) { do the work;}

/*static*/ bool CgroupV2Setup::CleanupCgroupV2ForContext(
    const PhysicalModeExecutionContext &ctx) { do work; }

...

#else

// all empty impls 

bool CgroupV2Setup::SetupCgroupV2ForContext(
    const PhysicalModeExecutionContext &ctx) {}
/*static*/ bool CgroupV2Setup::CleanupCgroupV2ForContext(
    const PhysicalModeExecutionContext &ctx){}

...

# endif
```

so we will only have 1 #ifdef in the whole file.

you may also want to put the util functions in the ifdef.

Reply:
Updated.